drivers/acpi/acpica/osl.c

```diff
@@ -66,4 +67,129 @@ ACPI_STATUS AcpiOsEnterSleep(UINT8 SleepState, UINT32 RegaValue, UINT32 RegbValu
 
     return AE_OK;
 }
+
+/* Memory and IO space read/write functions */
+
+ACPI_STATUS AcpiOsReadMemory(ACPI_PHYSICAL_ADDRESS Address, UINT64 *Value, UINT32 Width) {
+    void *pa = _ptr(_paddr(Address));
+    UINT64 val = 0;
+
+    switch (Width) {
+    case 8:
+        val = *(uint8_t *) pa;
+        break;
+    case 16:
+        val = *(uint16_t *) pa;
+        break;
+    case 32:
+        val = *(uint32_t *) pa;
+        break;
+    case 64:
+        val = *(uint64_t *) pa;
+        break;
+    default:
+        return AE_BAD_PARAMETER;
+    }
+
+    *Value = val;
+    return AE_OK;
+}
+
+ACPI_STATUS AcpiOsWriteMemory(ACPI_PHYSICAL_ADDRESS Address, UINT64 Value, UINT32 Width) {
+    void *pa = _ptr(_paddr(Address));
+
+    switch (Width) {
+    case 8:
+        *(uint8_t *) pa = (uint8_t) Value;
+        break;
+    case 16:
+        *(uint16_t *) pa = (uint16_t) Value;
+        break;
+    case 32:
+        *(uint32_t *) pa = (uint32_t) Value;
+        break;
+    case 64:
+        *(uint64_t *) pa = (uint64_t) Value;
+        break;
+    default:
+        return AE_BAD_PARAMETER;
+    }
+
+    return AE_OK;
+}
+
+ACPI_STATUS AcpiOsReadPort(ACPI_IO_ADDRESS Address, UINT32 *Value, UINT32 Width) {
+    switch (Width) {
+    case 8:
+        *Value = inb(Address);
+        break;
+    case 16:
+        *Value = inw(Address);
+        break;
+    case 32:
+        *Value = ind(Address);
+        break;
+    default:
+        return AE_BAD_PARAMETER;
+    }
+
+    return AE_OK;
+}
+
+ACPI_STATUS AcpiOsWritePort(ACPI_IO_ADDRESS Address, UINT32 Value, UINT32 Width) {
+    switch (Width) {
+    case 8:
+        outb(Address, (uint8_t) Value);
+        break;
+    case 16:
+        outw(Address, (uint16_t) Value);
+        break;
+    case 32:
+        outd(Address, (uint32_t) Value);
+        break;
+    default:
+        return AE_BAD_PARAMETER;
+    }
+
+    return AE_OK;
+}
+
+/* Memory management functions */
+
+void *AcpiOsAllocate(ACPI_SIZE Size) { return kmalloc(Size); }
+
+void AcpiOsFree(void *Memory) { kfree(Memory); }
+
+/* FIXME: Check if pages are mapped (with exception tables) */
+BOOLEAN AcpiOsReadable(void *Memory, ACPI_SIZE Length) { return true; }
```

Comment:
Until exception tables are implemented: can we check with frame->mapped?

Reply:
This is good idea. We should I think. But that would require finishing the implementation for vmm updating the mapped flag and adding more interfaces to PMM to find frame for a va. IMO it deserves a separate PR. Could you cut an issue for it please?